metriq_gym/benchmarks/chsh.py

```diff
@@ -0,0 +1,266 @@
+    """Constructs a subgraph of qubit pairs that violate the CHSH inequality.
+
+    Args:
+        job_data: The benchmark metadata including topology and coloring.
+        result_data: The result data containing measurement counts.
+
+    Returns:
+        The graph of edges that violated the CHSH inequality.
+    """
+    coloring = job_data.coloring
+
+    if coloring is None:
+        raise ValueError("Coloring information is required but not found in job_data.")
+
+    # A subgraph is constructed containing only the edges (qubit pairs) that successfully violate the CHSH inequality.
+    # The size of the largest connected component in this subgraph provides a measure of the device's performance.
+    good_edges = []
+    for job_idx, result in enumerate(result_data):
+        num_meas_pairs = len(
+            {key for key, val in coloring.edge_color_map.items() if val == job_idx}
+        )
+        exp_vals = np.zeros(num_meas_pairs, dtype=float)
+        for idx in range(4):
+            counts = result.measurement_counts[idx]
+            # Expectation values for the CHSH correlation terms are calculated based on measurement outcomes.
+            for pair in range(num_meas_pairs):
+                sub_counts = marginal_counts(counts, [2 * pair, 2 * pair + 1])
+                exp_val = sampled_expectation_value(sub_counts, "ZZ")
+                exp_vals[pair] += exp_val if idx != 2 else -exp_val
+        for idx, edge_idx in enumerate(
+            key for key, val in coloring.edge_color_map.items() if val == job_idx
+        ):
+            edge = (coloring.edge_index_map[edge_idx][0], coloring.edge_index_map[edge_idx][1])
+            # The benchmark checks whether the CHSH inequality is violated (i.e., the sum of correlations exceeds 2,
+            # indicating entanglement).
+            if exp_vals[idx] > 2:
+                good_edges.append(edge)
+    good_graph = rx.PyGraph(multigraph=False)
+    good_graph.add_nodes_from(list(range(coloring.num_nodes)))
+    for edge in good_edges:
+        good_graph.add_edge(*edge, 1)
+    return good_graph
+
+
+def largest_connected_size(good_graph: rx.PyGraph) -> int:
+    """Finds the size of the largest connected component in the CHSH subgraph.
+
+    Args:
+        good_graph: The graph of qubit pairs that violated CHSH inequality.
+
+    Returns:
+        The size of the largest connected component.
+    """
+    cc = rx.connected_components(good_graph)
+    largest_cc = cc[np.argmax([len(g) for g in cc])]
+    return len(largest_cc)
+
+
+class CHSH(Benchmark):
+    """Benchmark class for CHSH experiments."""
+
+    def dispatch_handler(self, device: QuantumDevice) -> CHSHData:
+        """Runs the benchmark and returns job metadata."""
+        shots = self.params.shots
+        topology_graph = None
+        coloring = None
+        trans_exp_sets = None
+
+        match device:
+            case QiskitBackend():
```

Comment:
I would branch by device only in one place in the code (you do it in the `device_coloring` function as well), so that it will be easier to extract it in the next iteration.

Reply:
Yep, localized these checks in one place here: fa23e6